include/downloader_pool.h

```diff
@@ -0,0 +1,248 @@
+#ifndef DOWNLOADER_POOL_H
+#define DOWNLOADER_POOL_H
+
+#include <optional>
+#include <thread>
+#include <functional>
+#include <map>
+#include <queue>
+#include <mutex>
+#include <vector>
+#include <list>
+
+#include "sockets.h"
+#include "http_downloader.h"
+#include "response.h"
+#include "uri.h"
+
+namespace Downloaders::Concurrency
+{
+	/// Callback that should accept HTTP response and should specify if download should be repeated
+	using FDownloadCallback = 
+		std::function<bool(std::optional<HTTP::CHTTPResponse>&&)>;
+
+	/// Special thread pool that carries conccurent HTTP download
+	/// provided by SocketClass that should inherit CSocket class
+	/// After each download from some URI running callback function
+	/// with passed response
+	template<class SocketClass>
+	class CConcurrentDownloader
+	{
+	static_assert(std::is_base_of<Sockets::CSocket, SocketClass>(), "SocketClass should inherit CSocket");
```

Comment:
Do you really need this check? Can the result of this check be different in this check?

Reply:
As I removed such check  from CHTTPDownloader in future, this one needs it, because it can construct wrong object